mango/models/genetic/population/base_population.py

```diff
@@ -0,0 +1,305 @@
+        elif self._selection_type == "rank":
+            self._elitism_selection()
+        elif self._selection_type == "random":
+            self._random_selection()
+        elif self._selection_type == "tournament":
+            self._tournament_selection()
+        else:
+            raise NotImplementedError("Selection method not implemented")
+
+    def crossover(self):
+        """
+        Method to run the crossover phase of the Genetic Algorithm
+
+        Currently, there is one crossover method implemented:
+            - Mask
+        """
+        if self._crossover_type == "mask":
+            self._mask_crossover()
+        else:
+            raise NotImplementedError("Crossover method not implemented")
+
+    def mutate(self):
+        """
+        Method to run the mutation phase of the Genetic Algorithm.
+
+        The actual implementation of the mutation is done in the individual as it is heavily dependent on the encoding.
+        """
+        for ind in self.offspring:
+            ind.mutate()
+
+    def update_population(self):
+        """
+        Method to make sure that all individuals in the population have a fitness value.
+        """
+        self._update_fitness(self.offspring + self.population)
+
+    def replace(self):
+        """
+        Method to run the replacement phase of the Genetic Algorithm.
+
+        Currently, there is one replacement method implemented:
+            - Elitist
+        """
+        if self._replacement_type == "elitist":
+            self._elitist_replacement()
+        else:
+            raise NotImplementedError("Replacement method not implemented")
+
+    def update_best(self):
+        """
+        Method to update the best adapted individual on the population.
+        """
+        if self._optimization_objective == "max":
+            current_best = max(self.population, key=lambda x: x.fitness)
+        else:
+            current_best = min(self.population, key=lambda x: x.fitness)
+
+        if self.best is None:
+            self.best = current_best
+        else:
+            if self._optimization_objective == "max":
+                if current_best.fitness > self.best.fitness:
+                    self.best = current_best
+            else:
+                if current_best.fitness < self.best.fitness:
+                    self.best = current_best
+
+    def stop(self):
+        """
+        Method to implement stop conditions based on information about the population or the genetic diversity.
+        """
+        # TODO: add more stop conditions
+        fitness_diversity = Counter([ind.fitness for ind in self.population])
+        if len(fitness_diversity) == 1:
+            print(
+                f"Exited due to low genetic diversity on generation: {self.generation+1}"
+            )
+            return True
+        if len(fitness_diversity) == 2:
+            if fitness_diversity.most_common(1)[0][1] == self.population_size - 1:
```

Comment:
There is a lot of "magic" index in this function can we define a more explaiinable variable for ```fitness_diversity.most_common(1)[0][1]```

Reply:
Will try to comment it or have it better